### ISSUE TITLE
Adjust SMW Factbox visibility

### ISSUE THREAD

**Issue by hypergrove** [SemanticMediaWiki/SemanticMediaWiki#975]
IN MW 1.25 and maybe earlier, there is a Page Information link that displays structured data associated with a page. The SMW Factbox shows on this Page Information page only when it's set to show on the page itself (per a flag). 

It seems a good idea that the SMW Factbox *always* show on a page's Page Information page (where action=info) and to apply the SMW Factbox flag only when action=view.

**Comment by Misdre** [comment]
Apparently SMW isn't using the InfoAction hook to add data to the action=info page... does that mean it's handled by OutputPageParserOutput?

**Comment by mwjames** [comment]
> Apparently SMW isn't using the InfoAction hook to add data to the action=info page

Correct, we don't use the hook because the `Factbox` existed long before `InfoAction` was introduced and I doubt that adding the hook brings any additional insight that is not already provided by the `Factbox` (as in-page inclusion) or by the `Special:Browse`. Unless we receive a PR (with appropriate tests), I can't see supporting the `InfoAction` to be happen in near future. 

> does that mean it's handled by OutputPageParserOutput?

https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/src/MediaWiki/Hooks/SkinAfterContent.php#L62

**Comment by hypergrove** [comment]
Agreed it's not a new function. This request stems from a desire to 
_conform_ to the addition of the InfoAction -- a new page that displays 
all structured data associated with a page.
SMW should not, in effect, be penalized for having developed its 
software prior to InfoAction being developed by the MW team.
I do request you reconsider your apparently dim view of this change request.
Cheers/john

On 4/29/2015 11:26 AM, mwjames wrote:
>
>     Apparently SMW isn't using the InfoAction hook to add data to the
>     action=info page
>
> Correct, we don't use the hook because the |Factbox| existed long 
> before |InfoAction| was introduced and I doubt that adding the hook 
> brings any additional insight that is not already provided by the 
> |Factbox| (as in-page inclusion) or by the |Special:Browse|. Unless we 
> receive a PR (with appropriate tests), I can't see supporting the 
> |InfoAction| to be happen in near future.
>
>     does that mean it's handled by OutputPageParserOutput?
>
> https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/src/MediaWiki/Hooks/SkinAfterContent.php#L62
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/SemanticMediaWiki/SemanticMediaWiki/issues/975#issuecomment-97532266>.
>

-- 
thanks/john
skype:hypergrove

**Comment by Misdre** [comment]
I don't think mwjames is against the change, he just does not see the value of supporting a new hook if the job is already taken care of by `SkinAfterContent`.

I'm trying to work on this but I have a lot of trouble just to reproduce the current behavior. It seems disabling the `Factbox` cache (`$smwgFactboxUseCache = false;`) disables the Factbox display entirely on `action=info` (it still works fine on the article page).

Steps to reproduce what I'm talking about:
1. **Factbox cache is enabled. Factbox display is enabled.**
    * `$smwgFactboxUseCache = true;` (default)
    * `$smwgShowFactbox = SMW_FACTBOX_NONEMPTY;`
2. Add some data to a random page.
    * `[[Is specified as::World Heritage Site]]`
3. The factbox is visible on the article page and on action=info.
4. Add a new annotation to the article.
    * `[[Has capital::Berlin]]`
5. The factbox is updated both on the article page and on action=info.


1. **Factbox cache is disabled. Factbox display is enabled.**
    * `$smwgFactboxUseCache = false;`
    * `$smwgShowFactbox = SMW_FACTBOX_NONEMPTY;`
2. Add some data to a random page.
    * `[[Is specified as::World Heritage Site]]`
3. The factbox is visible on the article page BUT NOT on action=info.
4. Add a new annotation to the article.
    * `[[Has capital::Berlin]]`
5. The factbox is updated on the article page but still not visible on action=info.

**Comment by Misdre** [comment]
Got it working (with the `InfoAction` hook), still have to write some tests/tidy up the code before raising a pull request.

**Comment by mwjames** [comment]
> Got it working (with the InfoAction hook), still have to write some tests/tidy up the code before raising a pull request.

Just to give you a heads up and not to discourage you, #1006 includes some minor name changes (things I did not get around before 2.2 rel).

> write some tests/tidy up the code before raising a pull request.

:+1: If you a question how to write a test just fire away or have a look at the `tests` folder (includes ~3000 of them and should give an idea how to do a test).

https://github.com/SemanticMediaWiki/SemanticMediaWiki/tree/master/tests#writing-tests

**Comment by Misdre** [comment]
Oh my... the "remove cruft" change is massive! Back to the drawing board I guess... Once I updated my code to work with your changes, can I open a pull request without the tests just for me to get some feedback?

**Comment by mwjames** [comment]
> Oh my... the "remove cruft" change is massive! 

Well, I couldn't make it smaller (it also shows the interdependencies that existed and the effort to make it go away).

> Back to the drawing board I guess... Once I updated my code to work 

Sorry for that but I had this in my branch for some time now and I just wanted it to disappear from my task list and not fight any more rebase battles.  

> I open a pull request without the tests just for me to get some feedback?

:+1:

**Comment by Misdre** [comment]
Well I probably don't need to open a pull request, since I pushed what I did to my own repo (see reference). It wasn't as hard to adapt it to your changes as I imagined it would be, and in fact the way the cache is working now is clearer.

About my code. I don't think there is anything hard to understand and it probably wasn't worth all the time I spent on it but hey, you gotta start somewhere. It comes down to `action=info` not calling the `OutputPageParserOutput` hook, thus never preparing the Factbox content before `SkinAfterContent`. If a cached version was available the factbox would be added to `action=info`, but disabling the cache would make it disappear.

So in  `InfoAction.php`, called on action=info and only there, I'm basically doing the `OutputPageParserOutput` job by preparing the factbox content, and then all work well (or seem to be).

Comments are welcome, don't be too harsh!
